tests/drivers/i2c/i2c_target_api/src/main.c

```diff
@@ -273,4 +276,47 @@ ZTEST(i2c_eeprom_target, test_eeprom_target)
 	}
 }
 
+ZTEST(i2c_eeprom_target, test_eeprom_2_read_only)
+{
+	const struct device *const eeprom_2 = DEVICE_DT_GET(NODE_EP2);
+	const struct device *const i2c_1 = DEVICE_DT_GET(DT_BUS(NODE_EP1));
+	int addr_2 = DT_REG_ADDR(NODE_EP2);
+	uint8_t addr_2_width = DT_PROP_OR(NODE_EP2, address_width, 8);
+	int ret;
+
+	/* Ensure the device and bus are ready */
+	zassert_not_null(i2c_1, "EEPROM 2 - I2C bus (i2c_1) not found");
+	zassert_not_null(eeprom_2, "EEPROM 2 device not found");
+	zassert_true(device_is_ready(i2c_1), "EEPROM 2 - I2C bus (i2c_1) not ready");
+
+	TC_PRINT("Found EEPROM 2 on I2C bus device %s at addr %02x\n", i2c_1->name, addr_2);
+
+	/* Step 1: Program EEPROM 2 with test data */
+	ret = eeprom_target_program(eeprom_2, eeprom_2_data, TEST_DATA_SIZE);
+	zassert_equal(ret, 0, "Failed to program EEPROM 2 with test data");
+
+	/* Step 2: Attach EEPROM 2 to its bus as a target device */
+	ret = i2c_target_driver_register(eeprom_2);
+	zassert_equal(ret, 0, "Failed to register EEPROM 2 as a target device");
+
+	/* Step 3: Verify EEPROM 2 data via I2C read */
+	ret = run_full_read(i2c_1, addr_2, addr_2_width, eeprom_2_data);
+	zassert_equal(ret, 0, "EEPROM 2 data verification via I2C read failed");
+
+	/* Step 4: Attempt to write to EEPROM 2 and expect a NACK or failure */
+	ret = i2c_write(i2c_1, write_attempt_data, TEST_DATA_SIZE, addr_2);
+
+	/* I2C_DW (and maybe others?) don't generate the NACK as expected */
+#ifndef CONFIG_I2C_DW
+	zassert_not_equal(ret, 0, "Write attempt to read-only EEPROM 2 succeeded unexpectedly");
+#endif
```

Comment:
Why is this needed?

Reply:
Usually the target should generate a NAK if it's being written and it's write protected, therefore the write on the controller side should return a non-zero return code. Unfortunately, the DesignWare I2C does not seem to implement this NAKing when used in target mode. I see if I can get hold of some other boards (e.g. STM32) to verify this as well.